docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+As this endpoint will be used for streaming large amounts of data it has the potential to be somewhat difficult to use as system limits could be pushed and errors could occur after annoyingly long time investments.  Some questions that may come up as useability becomes a bigger concern:
+
+* Is it possible that remote users pushing data to the endpoint could overload cluster's resources by pushing at a rate faster than processing can occur in cluster? If this is an issue how can the endpoint maintain connections in a resource-aware non-degrading way?  Will it be worthwhile to open up an endpoint running a different protocol (i.e. libp2p stream-muxer)?
+* What features can cluster provide to make the process easier?  For example if cluster could report an estimate aggregate of all of the space in its nodes' repos then users could check that a huge file will fit on the cluster before importing it.
+* What is the distribution of user preferences regarding pushing or pulling data to or into the endpoint?  Is it worth including location addressed link resolution (i.e. `ipfs-cluster-ctl add --shard http://www.awesomefilez.warez`), or file system device reading support?
+* Any others?
+
+#### Sharding existing DAGs
+An endpoint should also exist for sharding an existing DAG among an ipfs-cluster.  A user will run something like `ipfs-cluster-ctl pin --shard <cid>` and pin shards of the DAG rooted at `cid` across different nodes.  The importing step of the pipeline can be skipped in this case.  The naive implementation is to interact with the local daemon's DAG API to retrieve a stream of blocks as input to the sharding component.  We currently suspect that this is not going to work well for large DAGs that cannot fit in the node's repo.  The "Future Work" subsection: "sharding large existing DAGs" discusses some possibilities for scaling this to large DAGs.
+
+### Importer module
+
+As they are currently implemented in ipfs, importers run in three main phases.  The importer first begins by traversing a collection of data and metadata that can be interpreted as files and directories.  For example the tar importer reads data from a tar archive and the file importer used by `ipfs add` iterates over a collection of file objects.  Files are then broken into chunks with a chunker.  These chunks are then used as the data to make the dag-nodes constructed by the dag layout building process.  The dag-nodes are serialized as blocks and saved to the repo.
```

Comment:
I wish such clear concise explanations of how things work would be written way more often in our projects 👍